src/storage/access_observer.cpp

```diff
@@ -0,0 +1,21 @@
+#include "storage/access_observer.h"
+#include "storage/block_compactor.h"
+
+namespace terrier::storage {
+void AccessObserver::ObserveGCInvocation() {
+  gc_epoch_++;
+  for (auto it = last_touched_.begin(), end = last_touched_.end(); it != end;) {
+    if (it->second + COLD_DATA_EPOCH_THRESHOLD < gc_epoch_) {
+      compactor_->PutInQueue(it->first);
+      it = last_touched_.erase(it);
+    } else {
+      ++it;
+    }
+  }
+}
+
+void AccessObserver::ObserveWrite(RawBlock *block) {
+  if (block->insert_head_ == block->data_table_->GetBlockLayout().NumSlots()) last_touched_[block] = gc_epoch_;
```

Comment:
This conditional's purpose is non-obvious. You're checking that the block is full first to even be eligible to be tracked by the AccessObserver?